### ISSUE TITLE
Issues with Ctrl+F text finding in documents. 

### ISSUE THREAD

**Issue by axkibe** [texstudio-org/texstudio#965]
### Environment
- TeXstudio: Current Head ce6595446701aa6e9967a53d8281fa09ddfd56b1
- Qt: Using Qt Version 5.11.3, compiled with Qt 5.11.3 D
- OS: Debian buster
- TeX distribution: texlive

### Expected behavior
Ctrl+F text searching to work

### Actual behavior
Sometimes this doesn't work. I just got that as report, at first I considered it hard to believe but testing around I thought I experienced it myself, you search for a needle text, it says it's not in the document but it is. I supposed something was off with capitilasation search, or search by word, etc. Then it works again, and you know, then you doubt if you all imagined it only..?

Not sure I ran texstudio with "valgrind" and yo and behold I get tons of reports like this when Ctrl+F searching:
```
==8052== Conditional jump or move depends on uninitialised value(s)
==8052==    at 0x14D47178: ???
==8052==    by 0x19DAC087: ???
==8052== 
==8052== Conditional jump or move depends on uninitialised value(s)
==8052==    at 0x14D47178: ???
==8052==    by 0x19F63887: ???
==8052== 
==8052== Conditional jump or move depends on uninitialised value(s)
==8052==    at 0x14D47178: ???
==8052==    by 0x1A387C87: ???
==8052== 
==8052== Conditional jump or move depends on uninitialised value(s)
==8052==    at 0x14D47178: ???
==8052==    by 0x112101E7: ???
==8052== 
==8052== Conditional jump or move depends on uninitialised value(s)
==8052==    at 0x14D47178: ???
==8052==    by 0x13C50B67: ???
==8052== 
```
I did compile with debug symbols, don't know why gdb can't produce a backtrace.

I tried running valgrind as gdbserver and connecting to it... interesting in this mode, the unitionalized conditional reports vanish, but TeXStudio is now consistent in not finding anything when the needle is at least 3 letters long.

### How to reproduce
start TeXStudio like this:
$ valgrind /usr/local/bin/texstudio

Wait, open a file, hit Ctrl+F search for something that is clearly in the file, and see the reports of invalid memory usage.

Or start it in gdbserver mode:
valgrind --vgdb=yes --vgdb-error=0 /usr/local/bin/texstudio
-- another terminal --
gdb /usr/local/bin/texstudio 
target remote | vgdb

Now it won't find anything in the text file.

I see in gdb it starts new threads, can this be a race condition?

**Comment by sunderme** [comment]
the valgrind warning concern QT internal functions (Qt 5.14.1 here)
` ==22132== Conditional jump or move depends on uninitialised value(s)
==22132==    at 0x736EC7C: sljit_emit_cmov (in /home/sdm/Qt/5.14.1/gcc_64/lib/libQt5Core.so.5.14.1)
==22132==    by 0x736FB0E: mainloop_entry (in /home/sdm/Qt/5.14.1/gcc_64/lib/libQt5Core.so.5.14.1)
==22132==    by 0x738E98C: jit_compile (in /home/sdm/Qt/5.14.1/gcc_64/lib/libQt5Core.so.5.14.1)
`

Searching text works fine, though.

Please give a working example, i.e. the document plus what you are actually searching and from where.

**Comment by sunderme** [comment]
you could run `texstudio --execute-tests`  , searching is part of the auto-tests...

**Comment by MeanSquaredError** [comment]
Yes, Qt has these internal issues and a bunch of others too...

However searching works fine for me with the latest github head of TXS + Qt 5.12 and 5.13

The search is done in `QSearchReplacePanel::findReplace` which in turn invokes 
`if (!countOnly)  m_search->next(backward, replaceAll, !cbPrompt->isChecked(), true);`
`m_search->next` invokes 'QDocumentSearch::next()` which does the actual search.

So if you are able to reproduce the problem you can set a breakpoint on `QSearchReplacePanel::findReplace` and see what actually happens when it fails to find the search string. My guess is that there is some kind of memory corruption indeed, but it is difficult to fix it without reproducing it first.

**Comment by axkibe** [comment]
The issue is, I cannot reproduce it, but I've seen it. And I know how reproducing it, is a fundamental step to fixing it. That's why I got the idea it might happen to some memory corruption and tested with valgrind.

Two questions tough, Qt internal function or not, a conditional jump on uninitialized memory should never happen, right. So I'd report that to Qt?

Running valgrind in gdbserver mode... and search stops working, why tough?

**Comment by axkibe** [comment]
Okay I give up on reproducing it, maybe yesterday I was just too tired and made stupid mistakes...

**Comment by axkibe** [comment]
Maybe what may have been happening is that there was a space after the search string in the box, and thus it didn't find it anymore?

Technically it's correct, I see there may be a user friendliness quirk on it. I wouldn't know how to make it better tough

**Comment by MeanSquaredError** [comment]
1. Qt should not do conditional jumps on uninitialized memory. It is probably some bug in Qt, so it might be worth reporting it to Qt developers. From my experience they pay attention mostly to bug reports for the latest version of Qt, so it is probably worth reproducing it with the latest Qt 5.15.x before reporting it.

2. I wouldn't write off this search issue just as a user mistake. Actually I tried running valgrind/gdbserver+gdb and while search works fine, there is something fishy in the part where it handles search.
```
==2732== Invalid read of size 16
==2732==    at 0x253C92B5: ???
==2732==    by 0x254463CB: ???
==2732==  Address 0x254463ce is 30 bytes inside a block of size 40 alloc'd
==2732==    at 0x483980B: malloc (vg_replace_malloc.c:309)
==2732==    by 0x5D049C0: QArrayData::allocate(unsigned long, unsigned long, unsigned long, QFlags<QArrayData::AllocationOption>) (qarraydata.cpp:118)
==2732==    by 0x5D738D8: UnknownInlinedFun (qarraydata.h:224)
==2732==    by 0x5D738D8: QString::QString(QChar const*, int) (qstring.cpp:2153)
==2732==    by 0x5D77400: QString::left(int) const (qstring.cpp:5018)
==2732==    by 0xABF242: QDocumentSearch::next(bool, bool, bool, bool, QDocumentCursor const*) (qdocumentsearch.cpp:753)
==2732==    by 0xC157E3: QDocumentSearchTest::next_sameText() (qdocumentsearch_t.cpp:253)
==2732==    by 0xCCBB26: QDocumentSearchTest::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (moc_qdocumentsearch_t.cpp:99)
==2732==    by 0x5EB7C19: QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) const (qmetaobject.cpp:2308)
==2732==    by 0x5BFE779: invoke (qmetaobject.h:122)
==2732==    by 0x5BFE779: QTest::TestMethods::invokeTestOnData(int) const (qtestcase.cpp:941)
==2732==    by 0x5BFF0E2: QTest::TestMethods::invokeTest(int, char const*, QTest::WatchDog*) const (qtestcase.cpp:1140)
==2732==    by 0x5BFF6D0: QTest::TestMethods::invokeTests(QObject*) const (qtestcase.cpp:1484)
==2732==    by 0x5BFFABA: QTest::qRun() (qtestcase.cpp:1922)
```
I built TXS against Qt 5.13 (the Qt version used in Fedora 31). I strongly suspect that there is something buggy either in Qt or in TXS, so I am going to investigate it in more detail. It could be just a bug that is fixed in the latest Qt, but still it is worth investigating.

I will post my findings here once I check this issue in more detail.

**Comment by MeanSquaredError** [comment]
I can confirm that the valgrind warnings and errors are caused by bug(s) in `pcre2`, a library used by `QRegularExpression`.

On my computer their jit compiler builds the regular expression into machine code that uses 128-bit xmm registers. While executing the jit-compiled code these xmm registers use xmm instructions to read data from the string in which we search. Since the xmm registers are large (16-byte) they quite often read a few bytes past the end of the string, especially when we search close to the end of the string. This happens really easy in our tests when we do a backwards search that starts from the end of the string.

I will investigate it in more detail and then will file a bugreport with the `pcre2` library.

**Comment by axkibe** [comment]
Thanks for taking this seriously and investigating the issues.

However, I have another theory what is likely happening and which I actually observed once, by default the Option "restrict search scope to an existing selection" is enabled... and this ... an unusual experience compared to most other software, I believe thats why I got the complaints of TeXstudio not finding stuff that is certainly there. Turned it off, and "search work more reliable now"...

**Comment by MeanSquaredError** [comment]
OK, I finished investigating the issue with the Valgrind warnings when using QString regular expression search in TXS.

The TL;DR; is : These warnings are harmless and caused by highly-optimized machine code. People running TXS with Valgrind should  create a custom Valgrind suppression file to ignore these warnings.

The long story:

1. The Qt's QString uses internally the PCRE2 library which does all the regular expression matching.
2. When running on the x86 platform the PCRE2 library uses the so-called Just-In-Time (JIT) compilation of custom search code. Bascially each time when we search for some regular expression, the PCRE2 JIT compiler will compile on the fly a custom search function written in x86 machine code.
3. For efficiency reasons the compiled search code will make wide use of SSE/SIMD instructions whenever the x86 platform supports it.
4. Again for efficiency reasons the SSE/SIMD instructions read data is 16-byte blocks that are aligned on 16-byte boundaries. (actually these reads are aligned only in about 50% of the reads, but that is not important).
5. Since the reads are done in 16-byte, aligned blocks, quite often they read a few bytes before or after the QString data block that has been allocated through the heap manager or on the stack.
6. The JIT-compiled code knows that some of the read data falls outside of the actual subject string (the string in which we search). So it does a few additional checks when a match is found and if the match falls outside of the subject string range, that match is ignored and treated as if no match was found.
7. Thanks to the aligned 16-byte reads, the fact that each OS page is a multiple of 1024 bytes, and to the fact that the unaligned reads are carefully crafted, there is no danger (or at least should be no danger) of the JIT code trying to read inaccessible memory and therefore triggering an exception and crash.

Valgrind does not know that the reads outisde of the allocated buffer are safe and ignored afterwards, so Valgrind logs a number of warnings. There are two possible solutions:

I have discussed some of these issues with the PCRE2 maintainers here:
https://bugs.exim.org/show_bug.cgi?id=2540
and they provided some of the above information. Other information was discovered by me by actually disassembling the JIT code and spending some time digging into the custom compiled SIMD code and also into the PCRE2 JIT compiler. In the unlikely case that someone wants to spend the time and double-check these findings, I can provide the disassembly of the first JIT code that is compiled in TXS and that triggers a Valgrind warning. The disassembly is similar to the one that I posted on the PCRE2 bugtracker, but it has more comments that I made while digging into all the gory details of the JIT compiler.

Regarding the actual Valgrind warnings, the PCRE2 maintainers suggested two workarounds:

A) When debugging TXS with ValgrindCreate a custom suppression file for Valgrind can be written and used to ignore the warnings in the JIT code. This is my preferred solution since it does not require making any changes to the PCRE2 or Qt libraries.

B) PCRE2 can be configured and compiled with --enable-valgrind, then Qt should be linked with that custom build of PCRE2. After some digging on the PCRE2 bugtracker, I found some information that what --enable-valgrind does is to disable the JIT SIMD instructions. This second approach is definitely possible but not as good as the first approach because it requires rebuilding PCRE2 and Qt, it slows down QString regular expression search and additionally it provides different code paths for the TXS developer and the TXS users which may allow some of the TXS bugs to slip unnoticed by the developers.

**Comment by MeanSquaredError** [comment]
And by the way, when debugging the TXS/QString/PCRE2 under Valgrind+GDB it only reported correct addresses when Valgrind was run with `--vgdb=full`, i.e. the whole line was something like:
`valgrind --vgdb=full --vgdb-error=0 /usr/local/src/texstudio/texstudio`

**Comment by MeanSquaredError** [comment]
And also in the initial message you say: ``I did compile with debug symbols, don't know why gdb can't produce a backtrace.```
GDB does not show the backtrace because the Valgrind warnings happen in the JIT code that was compiled by PCRE2 itself. That code was not compiled when TXS was built, so it does not have debug symbols and GDB cannot show a backtrace for that JIT code.

**Comment by sunderme** [comment]
Kudos to you !
